python/lib/core/dmod/core/common/collections/bag.py

```diff
@@ -0,0 +1,126 @@
+"""
+Defines a collection of unordered data not requiring a hash
+"""
+from __future__ import annotations
+
+import typing
+
+from typing_extensions import Self
+
+
+_T = typing.TypeVar("_T")
+
+
+class Bag(typing.Collection[_T]):
+    """
+    A wrapper collection that hides functions/elements that treat the contents as anything other than an abstract
+    collection
+
+    Elements do not have to be hashable nor unique
+
+    Example Use Case:
+
+        You need to represent collected data that is meant to be unordered, but not unique/requiring a hash.
+        This leaves out list and set types.
+    """
+    def __init__(self, data: typing.Collection[_T] = None):
+        self.__data = [value for value in data] if data is not None else list()
+
+    def to_list(self) -> typing.List[_T]:
+        """
+        Convert the data into a normal list
+
+        Returns:
+            A list of the values within the bag
+        """
+        return [value for value in self.__data]
+
+    def add(self, value: _T) -> Self:
+        """
+        Add a value to the bag
+
+        Args:
+            value: The item to add
+
+        Returns:
+            The updated bag
+        """
+        self.__data.append(value)
+        return self
+
+    def find(self, condition: typing.Callable[[_T], bool]) -> typing.Optional[_T]:
+        """
+        Find the first item in the bag that matches the given condition
+
+        Args:
+            condition: A function defining if the encountered element counts as the one the caller is looking for
+
+        Returns:
+            The first item in the collection that matches the condition
+        """
+        for entry in self.__data:
+            if condition(entry):
+                return entry
+
+        return None
+
+    def filter(self, condition: typing.Callable[[_T], bool]) -> typing.Sequence[_T]:
+        """
+        Create a new collection containing only the elements that match the given condition
+
+        Args:
+            condition: A function defining what should appear within the new collection
+
+        Returns:
+            A new collection containing only elements matching the given condition
+        """
+        return [
+            entry
+            for entry in self.__data
+            if condition(entry)
+        ]
+
+    def remove(self, element: _T):
+        """
+        Remove an element from the bag
+
+        Args:
+            element: The element to remove
+        """
+        if element in self.__data:
+            self.__data.remove(element)
+
+    def pick(self) -> typing.Optional[_T]:
+        """
+        Extract an element from the bag
+
+        Returns:
+            An element from the bag if one exists
+        """
+        extracted_value: typing.Optional[_T] = None
+
+        if len(self.__data) > 0:
+            extracted_value = self.__data.pop()
+
+        return extracted_value
+
+    def count(self, element: _T) -> int:
```

Comment:
This function doesn't have a predicate attached to it since most implementations of `count` for lists/bags/multisets/whatever don't have predicates for this function.

Reply:
Yeah I think that is fair. Having a predicate would allow you to do some pretty powerful stuff in this instance though.